examples/aes-ccm/src/ccm_mode.rs

```diff
@@ -0,0 +1,210 @@
+    }
+
+    // creation of b(1) to b(u)
+    let mut k: usize = 16; // byte number to set next
+
+    if a_octets > 0 {
+        let x = 0xffu8;
+        let y = 0xfeu8;
+        let mut alen_copy = alen;
+
+        if alen >= 0x800000 {
+            b[16] = U8(x);
+            k = k + 2;
+
+            if alen < 0x100000000 {
+                b[17] = U8(y);
+            } else {
+                b[17] = U8(x);
+            }
+        }
+
+        for i in 1..a_octets+1 {
+            let curr = (alen_copy & 0xFF) as u8;
+            b[k + a_octets - i] = U8(curr);
+            alen_copy = alen_copy >> 8;
+        }
+
+        k = k + a_octets;
+
+        for i in 0..alen {
+            b[i+k]= a[i];
+        }
+
+        k = k + alen;
+
+        for _t in 0..16 {
+            if k % 16 != 0 {
+                // add zero padding for Associated Data
+                b[k] = U8(0u8);
+                k = k + 1;
+            }
+        }
+    }
+
+    // creation of b(u+1) to b(r)
+    for i in 0..plen {
+        b[i+k] = p[i];
+    }
+
+    k = k + plen;
+
+    for _t in 0..16 {
+        if k % 16 != 0 {
+            // add zero padding for Payload
+            b[k] = U8(0u8);
+            k = k + 1;
+        }
+    }
+
+    b
+}
+
+fn get_t(b: &ByteSeq, key: Key128, num: usize) -> ByteSeq {
+    let b0 = b.get_exact_chunk(16, 0);
+    let bloc = Block::from_seq(&b0);
+    let mut y_curr = aes128_encrypt_block(key, bloc);
+
+    for i in 1..b.len()/16 {
+        let mut b_curr = Block::from_seq(&b.get_exact_chunk(16, i));
+        b_curr = y_curr ^ b_curr;
+        y_curr = aes128_encrypt_block(key, b_curr);
+    }
+
+    ByteSeq::from_seq(&(y_curr.slice(0, num)))
+}
+
+fn counter_func(n: &ByteSeq, nlen: usize, m: usize) -> ByteSeq {
+    let mut ctr = ByteSeq::new(16 * (m + 1));
+    let qlen: u8 = 15u8 - (nlen as u8);
+    let flag = qlen - 1u8;
+
+    for i in 0..m+1 {
+        let mut icopy = i as u64;
+
+        let k = 16 * i;
+        ctr[k] = U8(flag);
+
+        for j in 0..nlen {
+            ctr[k+j+1] = n[j];
+        }
+
+        for x in 1..15-nlen {
+            let curr = icopy & 0xFFu64;
+            ctr[k+16-x] = U8(curr as u8);
+            icopy = icopy >> 8;
+        }
+    }
+
+    ctr
+}
+
+fn ctr_cipher(ctr: &ByteSeq, key: Key128, m: usize) -> (ByteSeq, ByteSeq) {
+    let ctr_zero = Block::from_seq(&ctr.get_exact_chunk(16, 0));
+    let key_copy = key.clone();
+    let s0 = ByteSeq::from_seq(&aes128_encrypt_block(key, ctr_zero));
+    let mut s = ByteSeq::new(16 * m);
+
+    for i in 1..m+1 {
+        let new_copy = key_copy.clone();
+        let ctr_block = Block::from_seq(&ctr.get_exact_chunk(16, i));
+        let s_curr = aes128_encrypt_block(new_copy, ctr_block);
+        let seq_s = ByteSeq::from_seq(&s_curr);
+        s = s.set_exact_chunk(16, i-1, &seq_s);
+    }
+
+    (s0, s)
+}
+
+pub fn encrypt_ccm(a: ByteSeq, n: ByteSeq, pay: ByteSeq, key: Key128, tlen: usize) -> ByteSeq {
+    let key_copy = key.clone();
```

Comment:
It doesn't work that way, because I'm using the key multiple times in a loop for encrypting each block one by one. I had tried many variations of the cloning for the `key` and the tests ran successfully but the typechecker failed for all my attempts other than the way I've done it right now.